frame/pov-limit/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Pov Limit Pallet
+//!
+//! Pallet that controls compute usage and PoV size usage. The parameters for
+//! the usage are configured by root and are stored as `StorageValue`s.
+//!
+//! NOTE: This is only meant to be used for testing.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+pub mod weights;
+
+use frame_support::{pallet_prelude::*, traits::GenesisBuild};
+use sp_core::{Blake2Hasher, Hasher};
+use sp_runtime::Perbill;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The number of times the hash function should be called to fill up
+		/// all the block's weight.
+		#[pallet::constant]
+		type HashesForFull: Get<u32>;
+
+		/// Weight information for this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub(crate) type Compute<T: Config> = StorageValue<_, Perbill, ValueQuery>;
+
+	#[pallet::storage]
+	pub(crate) type Storage<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig {
+		pub compute: Perbill,
+		pub storage: u32,
+	}
+
+	#[cfg(feature = "std")]
+	impl Default for GenesisConfig {
+		fn default() -> Self {
+			Self { compute: Default::default(), storage: Default::default() }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {
+			Compute::<T>::set(self.compute);
+			Storage::<T>::set(self.storage);
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+			let mut weight = T::DbWeight::get().reads(1);
+
+			for i in 0..(Compute::<T>::get().mul_ceil(T::HashesForFull::get())) {
+				weight.saturating_add(T::WeightInfo::hash_value());
+				if remaining_weight.any_lt(weight) {
+					weight = remaining_weight;
+					break
+				}
+				Self::hash_value(i.into());
+			}
+
+			for i in 0..Storage::<T>::get() {
+				weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+				if remaining_weight.any_lt(weight) {
+					weight = remaining_weight;
+					break
+				}
+
+				storage::unhashed::put(&i.to_le_bytes(), &i.to_le_bytes());
```

Comment:
Writing does not go into the PoV, only reading. This should also be caught by the test that I mentioned.  
Not sure what we can read though… maybe accounts? Or we could make it configurable with a trait.